installinator/src/peers.rs

```diff
@@ -0,0 +1,334 @@
+use buf_list::BufList;
+use bytes::{Bytes, BytesMut};
+use display_error_chain::DisplayErrorChain;
+use itertools::Itertools;
+use progenitor_client::ResponseValue;
+use reqwest::StatusCode;
+use tokio::{
+    sync::{mpsc, oneshot},
+    time::Instant,
+};
+
+use crate::errors::{ArtifactFetchError, DiscoverPeersError};
+
+/// A fetched artifact
+pub(crate) struct FetchedArtifact {
+    pub(crate) attempt: usize,
+    pub(crate) addr: Ipv6Addr,
+    pub(crate) artifact: BufList,
+}
+
+impl FetchedArtifact {
+    /// In a loop, discover peers, and fetch from them.
+    ///
+    /// This only produces an error if the discover function errors out. In normal use, it is expected
+    /// to never error out.
+    pub(crate) async fn loop_fetch_from_peers<F, Fut>(
+        log: &slog::Logger,
+        mut discover_fn: F,
+        artifact_id: &ArtifactId,
+    ) -> Result<Self>
+    where
+        F: FnMut() -> Fut,
+        Fut: Future<Output = Result<Peers, DiscoverPeersError>>,
+    {
+        let mut attempt = 0;
+        loop {
+            attempt += 1;
+            let peers = match discover_fn().await {
+                Ok(peers) => peers,
+                Err(DiscoverPeersError::Retry(error)) => {
+                    slog::debug!(
+                        log,
+                        "(attempt {attempt}) failed to discover peers, retrying: {}",
+                        DisplayErrorChain::new(AsRef::<dyn std::error::Error>::as_ref(&error)),
+                    );
+                    // XXX: add a delay here?
+                    continue;
+                }
+                Err(DiscoverPeersError::Abort(error)) => {
+                    return Err(error);
+                }
+            };
+
+            slog::debug!(
+                log,
+                "discovered {} peers: [{}]",
+                peers.peer_count(),
+                peers.display(),
+            );
+            match peers.fetch_artifact(artifact_id).await {
+                Some((addr, artifact)) => {
+                    return Ok(Self { attempt, addr, artifact })
+                }
+                None => {
+                    slog::debug!(
+                        log,
+                        "unable to fetch artifact from peers, retrying",
+                    );
+                }
+            }
+        }
+    }
+}
+
+impl fmt::Debug for FetchedArtifact {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FetchedArtifact")
+            .field("attempt", &self.attempt)
+            .field("addr", &self.addr)
+            .field(
+                "artifact",
+                &format!(
+                    "({} bytes in {} chunks)",
+                    self.artifact.num_bytes(),
+                    self.artifact.num_chunks()
+                ),
+            )
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Peers {
+    log: slog::Logger,
+    imp: Box<dyn PeersImpl>,
+    timeout: Duration,
+}
```

Comment:
I'm a little confused by the layering here.

We have:

- `struct Peers`: Looks... mostly real? but exposes some "mock" methods? But also *wraps* a `dyn` trait which could be used to inject a mock peer?
- `trait PeersImpl`: Seems like a decent spot to split between mock / not-mocked peers?
- `struct MockPeers` (in `peers.rs`): Contains a `Vec<Ipv6Addr>`, and I don't actually know how "mocked" this is supposed to be? There are comments inside about reaching out over the boostrap network?
- `struct MockPeers` (in `mock_peers.rs`): Contains a `BTreeMap<Ipv6Addr, MockPeer>`, and seems to be an actually mocked implementation?

What's the intended layering? Is one of the "MockPeers" supposed to be a "RealPeers"? In that case, is the layering supposed to be:
- "struct Peers" contains a "dyn PeersImpl" which might be a mock or real peer implementation?

Or something else? If that's the case, why is a `mock_discover` method being exposed from "struct Peers"? Shouldn't the "top-level" peers be agnostic to the realness / fakenss of the underlying peer impl?

Reply:
> "struct Peers" contains a "dyn PeersImpl" which might be a mock or real peer implementation?

Yeah, basically this.

> If that's the case, why is a mock_discover method being exposed from "struct Peers"? Shouldn't the "top-level" peers be agnostic to the realness / fakenss of the underlying peer impl?

I agree, will fix! (This code is all meant to be deleted anyway.)

Reply:
Fixed. I've also renamed the `MockPeers` in `peers.rs` to `ExamplePeers` -- the plan is for `ExamplePeers` to go away before this PR lands.